src/dsl.jl

```diff
@@ -517,98 +486,146 @@ function extract_species_and_parameters!(reactions, excluded_syms)
-    subs_stoich_init = deepcopy(subs_init)
-    prod_init = isempty(rxstruct.products) ? nothing : :([])
-    prod_stoich_init = deepcopy(prod_init)
-    reaction_func = :(Reaction($(recursive_expand_functions!(rxstruct.rate)), $subs_init,
-        $prod_init, $subs_stoich_init, $prod_stoich_init,
-        metadata = $(rxstruct.metadata)))
-    for sub in rxstruct.substrates
-        push!(reaction_func.args[3].args, sub.reactant)
-        push!(reaction_func.args[5].args, sub.stoichiometry)
-    end
-    for prod in rxstruct.products
-        push!(reaction_func.args[4].args, prod.reactant)
-        push!(reaction_func.args[6].args, prod.stoichiometry)
+    if haskey(options, :parameters)
+        pexprs = options[:parameters]
+    elseif isempty(parameters_extracted)
+        pexprs = :()
+    else
+        pexprs = :(@parameters)
     end
-    reaction_func
+    foreach(p -> push!(pexprs.args, p), parameters_extracted)
+    return pexprs
 end
 
-# takes a ModelingToolkit declaration macro like @parameters and returns an expression
-# that calls the macro and then scalarizes all the symbols created into a vector of Nums
-function scalarize_macro(nonempty, ex, name)
+# Takes a ModelingToolkit declaration macro (like @parameters ...) and return and expression:
+# That calls the macro and then scalarizes all the symbols created into a vector of Nums.
+# stores the created symbolic variables in a variable (which name is generated from `name`).
+# It will also return the name used for the variable that stores the symbolci variables.
+function scalarize_macro(expr_init, name)
+    # Generates a random variable name which (in generated code) will store the produced
+    # symbolic variables (e.g. `var"##ps#384"`).
     namesym = gensym(name)
-    if nonempty
+
+    # If the input expression is non-emtpy, wraps it with addiional information.
+    if expr_init != :(())
         symvec = gensym()
-        ex = quote
-            $symvec = $ex
+        expr = quote
+            $symvec = $expr_init
             $namesym = reduce(vcat, Symbolics.scalarize($symvec))
         end
     else
-        ex = :($namesym = Num[])
+        expr = :($namesym = Num[])
+    end
+
+    return expr, namesym
+end
+
+# From the system reactions (as `ReactionInternal`s) and equations (as expressions),
+# creates the expressions that evalutes to the reaction (+ equations) vector.
+function make_rxsexprs(reactions, equations)
+    rxsexprs = :(Catalyst.CatalystEqType[])
+    foreach(rx -> push!(rxsexprs.args, get_rxexpr(rx)), reactions)
+    foreach(eq -> push!(rxsexprs.args, eq), equations)
+    return rxsexprs
+end
+
+# From a `ReactionInternal` struct, creates the expression which evaluates to the creation
+# of the correponding reaction.
+function get_rxexpr(rx::ReactionInternal)
+    # Initiates the `Reaction` expression.
+    rate = recursive_expand_functions!(rx.rate)
+    rx_constructor = :(Reaction($rate, [], [], [], []; metadata = $(rx.metadata)))
```

Comment:
Please revert back to the old initialization. The point of using `nothing` was to avoid allocations. Creating empty arrays of type `Any` like here will result in unneeded allocations I believe (and inside the `Reaction` it will immediately set several of them to `nothing` anyways).